sshd-core/src/main/java/org/apache/sshd/common/kex/extension/DefaultClientKexExtensionHandler.java

```diff
@@ -52,247 +43,111 @@
  * session by adding the <A HREF="https://tools.ietf.org/html/rfc8332">&quot;rsa-sha2-256/512&quot;</A> signature
  * factories (if not already added).
  *
- * <B>Note:</B> experimental - used for development purposes and as an example
- *
  * @author <a href="mailto:dev@mina.apache.org">Apache MINA SSHD Project</a>
  */
 public class DefaultClientKexExtensionHandler extends AbstractLoggingBean implements KexExtensionHandler {
-    /**
-     * Session {@link AttributeKey} used to store the client's proposal
-     */
-    public static final AttributeKey<Map<KexProposalOption, String>> CLIENT_PROPOSAL_KEY = new AttributeKey<>();
+
+    /** Default singleton instance. */
+    public static final DefaultClientKexExtensionHandler INSTANCE = new DefaultClientKexExtensionHandler();
 
     /**
-     * Session {@link AttributeKey} used to store the server's proposal
+     * Session {@link AttributeKey} used to store whether the extension indicator was already sent.
      */
-    public static final AttributeKey<Map<KexProposalOption, String>> SERVER_PROPOSAL_KEY = new AttributeKey<>();
-
-    public static final NavigableSet<String> DEFAULT_EXTRA_SIGNATURES = Collections.unmodifiableNavigableSet(
-            GenericUtils.asSortedSet(String.CASE_INSENSITIVE_ORDER,
-                    KeyUtils.RSA_SHA256_KEY_TYPE_ALIAS,
-                    KeyUtils.RSA_SHA512_KEY_TYPE_ALIAS));
-
-    public static final DefaultClientKexExtensionHandler INSTANCE = new DefaultClientKexExtensionHandler();
+    private static final AttributeKey<Boolean> CLIENT_PROPOSAL_MADE = new AttributeKey<>();
```

Comment:
I don't see why this should be `private` - it is immutable and someone might find it useful. In general, I try not to "hide" anything from the users (within reason) - even if they can cause harm by misuse - since one never knows how our users might need to extend/our code.

Reply:
I know we have different ideas about interfaces. Not the first time this issue crops up :-) If you really think this should be accessible, let's add an accessor method.

If this is public, a client could reset this -- with might break rekeying? According to RFC 8308 the indicator is to be added only on the _first_ key exchange.

Reply:
>> I know we have different ideas about interfaces. Not the first time this issue crops up :-)
Indeed :-)

>> If you really think this should be accessible, let's add an accessor method

How would this prevent misuse ?

>> If this is public, a client could reset this -- with might break rekeying? According to RFC 8308 the indicator is to be added only on the first key exchange.

Like I said - I prefer holding the users responsible for misuse rather than patronize them and "protect them from themselves"

Reply:
Done, though I'd still have preferred at least a method like `public boolean wasClientExtensionProposalMade(Session session)` instead of a public attribute. We don't have to make it easy for clients to shoot their own feet.

Reply:
>> Done, though I'd still have preferred at least a method like public boolean wasClientExtensionProposalMade(Session session) instead of a public attribute. We don't have to make it easy for clients to shoot their own feet.

Like you said - we differ in our approach/view on this matter :-))

Reply:
Do you have commit rights to _master_ or do you need me to merge your PRs ? If so, let me know when you think it is ready (since I have no more comments) and I will do so.

Reply:
I should have, soon. Finally got around to enable 2FA here on Github, accounts are linked, so now I'm waiting for the organizational invite. If I understood the process correctly, that should then also enable commit rights on master.

Reply:
>> I'm waiting for the organizational invite

Note that the _master_ I am speaking about is not the _github_ one but rather the ASF one. Github is only a **mirror** of the ASF one - we never push to github...

Reply:
I had asked Guillaume a while ago about where I should commit/merge; he said he just pushed to the Github repo and that worked for him. There's also a number of commits at gitbox that are PR merges. So I gathered that indeed simply merging PRs the usual Github way would work.

Reply:
>> I had asked Guillaume a while ago about where I should commit/merge

Please ask him then how to merge the 2 PRs - in any case, if you need me to merge them let me know